File: Interviews/LeetCode/Daily-Problems/2022/November/Day-22/279.cpp
class Solution {
   public:
    int numSquares(int n) {
        int sq = (int)sqrt(n);
        vector<int> cache(n + 1, 0);
        for (int i = 1; i <= n; i++) {
            cache[i] = 1e9;
            for (int j = 1; j <= sq; j++) {
                if (j * j <= i) {
                    cache[i] = min(cache[i], cache[i - j * j] + 1);
                }
            }
        }

        return cache[n];
    }
};

File: Interviews/LeetCode/Daily-Problems/2022/November/Day-22/279.py
class Solution:
    def numSquares(self, n: int) -> int:
        sq = math.isqrt(n)
        cache = [0 for i in range(n + 1)]
        for i in range(1, n + 1):
            cache[i] = int(1e9)
            for j in range(1, sq + 1):
                if j * j <= i:
                    cache[i] = min(cache[i], cache[i - j * j] + 1)

        return cache[n]

# This solution TLEs on LeetCode. The same solution as the C++ version.
